### PR TITLE
feat: workflow abstraction and parallelize Join Parts

### PR DESCRIPTION
**Pull request by donghanz**
## Summary

- abstract join backfill workflow, so that it can be easily integrated with different orchestrators 
- update backfill workflow, so that join part computation is parallelized
- implement Airflow orchestrator: input Flow object and output the Airflow DAG for the join backfill
- helper methods
- add `isort` and `autoflake` to pre-commit

## Next steps
- Find a better way to read configs in the Airflow DAG (for example, read from S3)
- Implement `run_left()` and `run_final_join` once the Spark side changes are done
- Integrate `join_backfill` with `run.py`

## Test Plan

- [x] Added Unit Tests
- [x] Covered by existing CI
- [x] Integration tested: https://airflow-stone.d.musta.ch/tree?base_date=&num_runs=1&dag_id=join_backfill&arrange=UD&root=
image

## Checklist
- [] Documentation update: will update later in another PR

## Reviewers
@nikhilsimha @ezvz @better365

### CODE REVIEW

File: api/py/ai/chronon/repo/join_backfill.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+from ai.chronon.utils import join_part_name
+
+SPARK_VERSION = "3.1.1"
+SPARK_JAR_TYPE = "uber"
+logging.basicConfig(level=logging.INFO)
+airflow_client.init(airflow_client.Service.STONE)
+
+
+def test():
+    pass
```
Comment (better365) [inline api/py/ai/chronon/repo/join_backfill.py:19]: the purpose of this placeholder method? remove it?

File: api/py/ai/chronon/scheduler/adapters/airflow_adapter.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+from ai.chronon.scheduler.interfaces.orchestrator import WorkflowOrchestrator
+
+from airflow import DAG
+from airflow.operators.bash_operator import BashOperator
+
+
+class AirflowOrchestrator(WorkflowOrchestrator):
```
Comment (better365) [inline api/py/ai/chronon/scheduler/adapters/airflow_adapter.py:7]: @ezvz Are we planning to introduce Airflow as a dependency?

File: api/py/ai/chronon/repo/join_backfill.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            jp_full_name = join_part_name(join_part)
+            jp_node = Node(jp_full_name, self.run_join_part(jp_full_name))
+            flow.add_node(jp_node)
+            jp_node.add_dependency(left_node)
+            final_node.add_dependency(jp_node)
+        return flow
+
+    def run_join_part(self, join_part: str):
+        # TODO: Find a better way to sync configs
+        cmd = "aws s3 cp s3://airbnb-datainfra-dependencies-internal-only/testing/test_online_join_small.v2 . && "
```
Comment (better365) [inline api/py/ai/chronon/repo/join_backfill.py:60]: let's remove Airbnb S3 bucket name and pass it as a param

File: api/py/ai/chronon/scheduler/adapters/airflow_adapter.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            schedule_interval=schedule_interval,
+        )
+
+    def schedule_task(self, node):
+        return BashOperator(task_id=node.name, dag=self.dag, bash_command=node.command)
+
+    def set_dependencies(self, task, dependencies):
+        task.set_upstream(dependencies)
+
+    def build_dag_from_flow(self, flow):
```
Comment (better365) [inline api/py/ai/chronon/scheduler/adapters/airflow_adapter.py:21]: This is very clean.

File: api/py/ai/chronon/scheduler/interfaces/flow.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+class Flow:
```
Comment (better365) [inline api/py/ai/chronon/scheduler/interfaces/flow.py:1]: Add a comment on the purpose the `Flow`?

File: api/py/ai/chronon/repo/run.py [hunk 6]
```diff
@@ -182,7 +185,13 @@ def download_only_once(url, path):
 
 
 @retry_decorator(retries=3, backoff=50)
-def download_jar(version, jar_type="uber", release_tag=None, spark_version="2.4.0"):
+def download_jar(
+    version,
+    jar_type="uber",
+    release_tag=None,
+    spark_version="2.4.0",
```
Comment (better365) [inline api/py/ai/chronon/repo/run.py:192]: we should be ready to use 3.1.1 as default now.

File: api/py/ai/chronon/scheduler/adapters/airflow_adapter.py (hunk 2 above)
Comment (donghanz) [reply api/py/ai/chronon/scheduler/adapters/airflow_adapter.py:7]: We will introduce Airflow but maintain orchestrator agnosticism by abstracting flow logic. Here is why we adopt Airflow: [doc](https://docs.google.com/document/d/1WF7WNAHegJfy_OH2Uux1eKf7avLDQRiuZ-nVWMKOLeo/edit#heading=h.cz7shfyfzg0n)

File: api/py/ai/chronon/repo/join_backfill.py (hunk 1 above)
Comment (donghanz) [reply api/py/ai/chronon/repo/join_backfill.py:19]: This is a placeholder for backfill dry run or for prototyping purpose. I can remove it for now and implement the detailed logic later.

File: api/py/ai/chronon/scheduler/interfaces/flow.py (hunk 5 above)
Comment (donghanz) [reply api/py/ai/chronon/scheduler/interfaces/flow.py:1]: Added

**Review by better365 — APPROVED** [review]

File: api/py/ai/chronon/utils.py [hunk 7]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (ezvz) [inline api/py/ai/chronon/utils.py:1]: Seems like these changes are mostly linting, so not reviewing them too close. Let me know if there's actual logic changes here to TAL at.

File: api/py/test/test_run.py [hunk 8]
```diff
@@ -16,22 +16,33 @@
 #     See the License for the specific language governing permissions and
 #     limitations under the License.
 
-from ai.chronon.repo import run
```
Comment (ezvz) [inline api/py/test/test_run.py:19]: Is this file also mostly linter changes?

**Review by ezvz — COMMENTED** [review]
Looks really good so far, @donghanz ! 

My main comment is just that I think we should still call run.py, or at least use the same logic/helper functions for executing Spark jobs.

File: api/py/ai/chronon/repo/join_backfill.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        for join_part in self.join.joinParts:
+            jp_full_name = join_part_name(join_part)
+            jp_node = Node(jp_full_name, self.run_join_part(jp_full_name))
+            flow.add_node(jp_node)
+            jp_node.add_dependency(left_node)
+            final_node.add_dependency(jp_node)
+        return flow
+
+    def run_join_part(self, join_part: str):
+        # TODO: Find a better way to sync configs
```
[the remark this answers is not in the document]
Comment (donghanz) [reply api/py/ai/chronon/repo/join_backfill.py:56]: Good point. Updated.

File: api/py/ai/chronon/scheduler/interfaces/orchestrator.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+from abc import ABC, abstractmethod
+
+
+class WorkflowOrchestrator(ABC):
```
[the remark this answers is not in the document]
Comment (donghanz) [reply api/py/ai/chronon/scheduler/interfaces/orchestrator.py:4]: Yes - it is the interface for all adapters/orchestrators.

File: api/py/ai/chronon/utils.py (hunk 7 above)
Comment (donghanz) [reply api/py/ai/chronon/utils.py:1]: `join_part_name()` is where the logic was updated. The rest is linting.

File: api/py/test/test_run.py (hunk 8 above)
Comment (donghanz) [reply api/py/test/test_run.py:19]: We added a new parameter for `download_jar()` and updated the unit tests accordingly in this file. For example: `def mock_cmd(url, path, skip_download):`

File: api/py/ai/chronon/repo/join_backfill.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+    def run_left(self):
+        # TODO: integrate with the Spark side change
+        return "echo 'Running left table'"
+
+    def run_final_join(self):
+        # TODO: integrate with the Spark side change
+        return "echo 'Running final join'"
+
+    def run(self, orchestrator: str):
```
Comment (donghanz) [inline api/py/ai/chronon/repo/join_backfill.py:97]: @hzding621 Airflow logic is extracted away, so JoinBackfill is independent from orchestrators now.

**Review by ezvz — APPROVED** [review]
Looks good!

**Pull request merged**